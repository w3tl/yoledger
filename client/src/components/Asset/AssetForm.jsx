import React from 'react';
import PropTypes from 'prop-types';
import { TextInput, Button } from '../Elements';
import AmountInput from '../AmountInput';
import { assetPropType } from './propTypes';

class AssetForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: props.asset.name,
      balance: props.asset.balance,
    };
  }

  handleChange = name => ({ target }) => {
    this.setState({
      [name]: target.value,
    });
  }

  handleSubmit = (e) => {
    e.preventDefault();
    const { name, balance } = this.state;
    const { onSave } = this.props;
    onSave({
      name,
      balance,
    });
  }

  render() {
    const { balance, name } = this.state;
    return (
      <form onSubmit={this.handleSubmit}>
        <TextInput label="Name" value={name} onChange={this.handleChange('name')} />
        <AmountInput label="Balance" value={balance} onChange={this.handleChange('balance')} />
        <Button type="submit">
          Save
        </Button>
      </form>
    );
  }
}

AssetForm.propTypes = {
  asset: assetPropType,
  onSave: PropTypes.func,
};
AssetForm.defaultProps = {
  asset: {
    id: null,
    name: '',
    balance: 0,
  },
  onSave: () => {},
};

export default AssetForm;
