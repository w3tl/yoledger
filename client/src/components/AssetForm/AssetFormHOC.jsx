import React from 'react';
import PropTypes from 'prop-types';

const assets = [{
  id: '1',
  name: 'Bank Card',
}, {
  id: '2',
  name: 'Cash',
}, {
  id: '3',
  name: 'Bank Card 2',
}];

const AssetFormWithData = (Component) => {
  const wrapped = (props) => {
    const { match: { params: { id } } } = props;
    const asset = assets.find(a => a.id === id);
    return <Component {...asset} {...props} />;
  };

  wrapped.propTypes = {
    match: PropTypes.shape({
      match: PropTypes.shape({
        params: PropTypes.shape({
          id: PropTypes.string,
        }),
      }),
    }),
  };
  wrapped.defaultProps = {
    match: {
      params: {
        id: null,
      },
    },
  };
  return wrapped;
};

AssetFormWithData.propTypes = {
  Component: PropTypes.element,
};

export default AssetFormWithData;
