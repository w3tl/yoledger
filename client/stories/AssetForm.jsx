import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import AssetForm from '../src/components/Asset/AssetForm';

const asset = {
  id: '1',
  name: 'Bank Card',
  balance: 0,
};

storiesOf('Asset/Form', module)
  .add('create', () => <AssetForm />)
  .add('edit', () => <AssetForm asset={asset} onSave={action('onSave')} />);
