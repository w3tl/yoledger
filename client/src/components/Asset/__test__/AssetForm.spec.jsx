import React from 'react';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import AssetForm from '../AssetForm';

configure({ adapter: new Adapter() });

const asset = {
  __typename: 'Account',
  id: '1',
  name: 'Card',
  balance: 15,
};

describe('AssetForm', () => {
  it('should render correctly', () => {
    const output = shallow(
      <AssetForm asset={asset} />,
    );
    expect(output.debug()).toMatchSnapshot();
  });

  it('should handle submit on edit mode', () => {
    const onSave = jest.fn();
    const output = mount(
      <AssetForm asset={asset} onSave={onSave} />,
    );
    output.find('form').simulate('submit');
    expect(onSave).toHaveBeenCalledWith({
      name: 'Card',
      balance: 15,
    });
  });

  it('should handle submit on create mode', () => {
    const onSave = jest.fn();
    const output = mount(
      <AssetForm onSave={onSave} />,
    );
    output.find('form').simulate('submit');
    expect(onSave).toHaveBeenCalledWith({
      name: '',
      balance: 0,
    });
  });
});
