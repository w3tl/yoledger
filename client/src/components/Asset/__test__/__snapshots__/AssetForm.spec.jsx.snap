// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetForm should render correctly 1`] = `
"<form onSubmit={[Function]}>
  <Component label=\\"Name\\" value=\\"Card\\" onChange={[Function]} />
  <Component label=\\"Balance\\" value={15} onChange={[Function]} />
  <Button type=\\"submit\\">
    Save
  </Button>
</form>"
`;
